brave/src/main/java/brave/propagation/CorrelationFieldScopeDecorator.java

```diff
@@ -0,0 +1,219 @@
+    ));
+
+    Builder(CorrelationField.Context context) {
+      if (context == null) throw new NullPointerException("context == null");
+      this.context = context;
+    }
+
+    /**
+     * Invoke this to clear fields so that you can {@linkplain #addField(CorrelationField) add the
+     * ones you need}.
+     *
+     * <p>Defaults may include a field you aren't using, such as "parentId". For better
+     * performance, only include the fields you use in your correlation expressions (such as log
+     * formats).
+     *
+     * @since 5.11
+     */
+    public Builder clearFields() {
+      this.fields.clear();
+      return this;
+    }
+
+    /** @since 5.11 */
+    public Builder addField(CorrelationField field) {
+      if (field == null) throw new NullPointerException("field == null");
+      if (field.name() == null) throw new NullPointerException("field.name() == null");
+      if (field.name().isEmpty()) throw new NullPointerException("field.name() isEmpty");
+      fields.add(field);
+      return this;
+    }
+
+    /** @throws IllegalArgumentException if no correlation fields were added. */
+    public final CorrelationFieldScopeDecorator build() {
+      int fieldCount = fields.size();
+      if (fieldCount == 0) throw new IllegalArgumentException("no correlation fields");
+      if (fieldCount == 1) {
+        return new Single(context, fields.iterator().next());
+      }
+      return new Multiple(context, fields);
+    }
+  }
+
+  final CorrelationField.Context context;
+
+  CorrelationFieldScopeDecorator(CorrelationField.Context context) {
+    this.context = context;
+  }
+
+  boolean update(@Nullable TraceContext traceContext, CorrelationField field,
+    String previousValue) {
+    String currentValue = traceContext != null ? field.getValue(traceContext) : null;
+    if (currentValue != null) {
+      if (!currentValue.equals(previousValue)) {
+        context.put(field.name(), currentValue);
+        return true;
+      }
+    } else if (previousValue != null) {
+      context.remove(field.name());
+      return true;
+    }
+    return false;
+  }
+
+  void replace(String name, @Nullable String value) {
+    if (value != null) {
+      context.put(name, value);
+    } else {
+      context.remove(name);
+    }
+  }
+
+  static final class Single extends CorrelationFieldScopeDecorator {
+    final CorrelationField field;
+
+    Single(CorrelationField.Context context, CorrelationField field) {
+      super(context);
+      this.field = field;
+    }
+
+    @Override public Scope decorateScope(TraceContext traceContext, Scope scope) {
+      if (scope == Scope.NOOP) return scope;
+
+      String previousValue = context.get(field.name());
+      if (!update(traceContext, field, previousValue)) {
+        return scope;
+      }
+
+      class SingleCorrelationFieldScope implements Scope {
+        @Override public void close() {
+          scope.close();
+          replace(field.name(), previousValue);
+        }
+      }
```

Comment:
made a comment elsewhere, but to read back the current value is expensive. also we allow late updates. finally we do want to clobber out of band updates because MDC can be used externally